Classes/Projection/ProjectionManager.php

```diff
@@ -0,0 +1,238 @@
+        $projectionsByName = $projectorClassNamesByIdentifier = [];
+        foreach ($reflectionService->getAllImplementationClassNamesForInterface(ProjectorInterface::class) as $projectorClassName) {
+            $package = $packageManager->getPackageByClassName($projectorClassName);
+            $projectionName = (new ClassReflection($projectorClassName))->getShortName();
+            if (substr($projectionName, -9) === 'Projector') {
+                $projectionName = substr($projectionName, 0, -9);
+            }
+            $projectionName = strtolower($projectionName);
+            $packageKey = strtolower($package->getPackageKey());
+            $projectionIdentifier = $packageKey . ':' . $projectionName;
+            if (isset($projectorClassNamesByIdentifier[$projectionIdentifier])) {
+                throw new \RuntimeException(sprintf('The projection identifier "%s" is ambiguous, please rename one of the classes "%s" or "%s"', $projectionIdentifier, $projectorClassNamesByIdentifier[$projectionIdentifier], $projectorClassName), 1476198478);
+            }
+            $projectorClassNamesByIdentifier[$projectionIdentifier] = $projectorClassName;
+            if (!isset($projectionsByName[$projectionName])) {
+                $projectionsByName[$projectionName] = [];
+            }
+            $projectionsByName[$projectionName][] = $packageKey;
+        }
+        ksort($projectorClassNamesByIdentifier);
+
+        $projections = [];
+        foreach ($projectorClassNamesByIdentifier as $fullProjectionIdentifier => $projectorClassName) {
+            list($packageKey, $projectionName) = explode(':', $fullProjectionIdentifier);
+            if (count($projectionsByName[$projectionName]) === 1) {
+                $shortIdentifier = $projectionName;
+            } else {
+                $shortIdentifier = $fullProjectionIdentifier;
+                $prefix = null;
+                foreach (array_reverse(explode('.', $packageKey)) as $packageKeyPart) {
+                    $prefix = $prefix === null ? $packageKeyPart : $packageKeyPart . '.' . $prefix;
+                    $matchingPackageKeys = array_filter($projectionsByName[$projectionName], function ($searchedPackageKey) use ($packageKey) {
+                        return $searchedPackageKey === $packageKey || substr($packageKey, -(strlen($searchedPackageKey) + 1)) === '.' . $searchedPackageKey;
+                    });
+                    if (count($matchingPackageKeys) === 1) {
+                        $shortIdentifier = $prefix . ':' . $projectionName;
+                        break;
+                    }
+                }
+            }
+            $projectionEventTypes = $eventListenerLocator->getEventTypesByListenerClassName($projectorClassName);
+            $projections[$fullProjectionIdentifier] = new Projection($fullProjectionIdentifier, $shortIdentifier, $projectorClassName, $projectionEventTypes);
+        }
```

Comment:
I'm still in favor of only supporting two projection identifier writings, `acme.foo.bar:quux` (full) or `quux` (shortest). If `quux` is given and is ambiguous, an exception is thrown with suggestions for the possible fully qualified projection names... this would be very easy to implement. KiSS FTW!

This whole method is just a big cryptic mess currently and even if it's only a very internal detail, I don't like it. I doubt I would ever have the guts to change anything in this method if needed, but it's fundamental for the whole projection replaying.

Sorry for the harsh feedback here, but I really don't see any real benefit but a lot of hard to follow code. Don't take it personally, I highly appreciate your massive efforts on this!